Prevents disassociate hybrid node on controlplane instance group

AlexSCorey opened the pull request:
##### SUMMARY
Addresses #11713. Prevents the user from being able to disassociate a hybrid node from the control plane instance group. In the UI inside Instance Groups --> controlplane instance group --> instances select hybrid instance.  Disassociate button should be disabled.

Then, click on the hybrid node.  In the details view the disassociate button should also be disabled.
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - API
 - UI

AlanCoding commented:
Why only hybrid nodes? That's how the issue was reported, but that's not how I thought the fix would go. It doesn't make sense to disassociate a "control" node from the controlplane either, and execution nodes shouldn't be associated to begin with. I think you should be able to drop that condition.

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -405,11 +405,23 @@ class InstanceInstanceGroupsList(InstanceGroupMembershipMixin, SubListCreateAtta
     def is_valid_relation(self, parent, sub, created=False):
         if parent.node_type == 'control':
-            return {'msg': _(f"Cannot change instance group membership of control-only node: {parent.hostname}.")}
+            return {'msg': _(f"Cannot change instance group membership of control-only node <<<<<<<<<<<<<<<<<<<<<<<<: {parent.hostname}.")}
```
looks like git conflict artifacts.

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -405,11 +405,23 @@ class InstanceInstanceGroupsList(InstanceGroupMembershipMixin, SubListCreateAtta
+        sub_id, res = super().unattach_validate(request)
+        if res:
+            return (sub_id, res)
+        parent = self.get_parent_object()
```
I don't think there's any need to obtain `parent` at all. All of these queries will be duplicates too.

AlanCoding replied:
This is why I suggested overriding `is_valid_relation`

nixocio commented:
hey @AlexSCorey as we discussed, see PR https://github.com/ansible/awx/pull/11839 may have some overlapping with the work on this PR.

AlexSCorey committed:
Disallows disassociate of hubrid type instances from controlplane instance group

Introduce new pattern for is_valid_removal

Makes disassociate error message a bit more dynamic

nixocio commented on awx/ui/src/screens/InstanceGroup/InstanceDetails/InstanceDetails.js:
```diff
@@ -275,10 +275,13 @@ function InstanceDetails({ setBreadcrumb, instanceGroup }) {
             <DisassociateButton
-              verifyCannotDisassociate={!me.is_superuser}
+              verifyCannotDisassociate={
+                !me.is_superuser || instanceGroup.name === 'controlplane'
```
From line 276, we just show this button if it is a super_user. We could simplify.
```suggestion
               instanceGroup.name === 'controlplane'
```

nixocio commented on awx/ui/src/screens/InstanceGroup/Instances/InstanceList.js:
```diff
@@ -224,13 +224,15 @@ function InstanceList() {
-                )}
+                verifyCannotDisassociate={
+                  selected.some((s) => s.node_type === 'control') ||
+                  instanceGroup.name === 'controlplane'
```
Do we need to verify 'hybrid' as well? To be similar to be other related logic. Did I misread this?

https://github.com/ansible/awx/blob/53a2679e3a77d90ed0f64ee38d2b2824e4751bba/awx/ui/src/components/DisassociateButton/DisassociateButton.js#L43

AlexSCorey replied:
We only need to prevent disassociation of an hybrid instance if the instance group it belongs to is a protected instance group.  IIRC my thought process was to put that logic here so as to help keep the logic in the instancegroup/instances/instancelist from being too complicated.

nixocio replied:
Thanks for clarifying @AlexSCorey

nixocio reviewed (commented):
@AlexSCorey, good job. I left a couple of questions related to UI changes.

nixocio reviewed (approved):
Approve the UI changes.

akus062381 commented:
can be merged as soon as @AlanCoding approves API changes :-)

AlanCoding commented:
I understand why @AlexSCorey did everything he did in this PR, but I'm not happy with the direction it moves the API in. I suggest another change in https://github.com/AlexSCorey/awx/pull/22

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -405,10 +405,8 @@ class InstanceInstanceGroupsList(InstanceGroupMembershipMixin, SubListCreateAtta
-        if parent.node_type == 'hop':
-            return {'msg': _(f"Cannot change instance group membership of hop node: {parent.hostname}.")}
+        if parent.node_type == 'control' or parent.node_type == 'hop' or (sub.name == 'controlplane' and parent.node_type == 'hybrid'):
+            return {'msg': _(f"Cannot change instance group membership of {parent.node_type} node: {parent.hostname}.")}
```
This message will say the wrong thing for hybrid nodes. You _can_ change memberships of hybrid nodes, but not for the control plane. @jbradberry along the lines of your recent comments.

The sneaky thing here is that instance-instance-group endpoints use `is_valid_relation` symmetrically, on attach and unattach, and no other types of views do this.

It may still be possible to resolve the issue with this approach, but need the wording to be right, and use the setting instead of hard-coding "controlplane"

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -506,11 +514,19 @@ class InstanceGroupInstanceList(InstanceGroupMembershipMixin, SubListAttachDetac
     search_fields = ('hostname',)
 
-    def is_valid_relation(self, parent, sub, created=False):
+    def is_valid_relation(self, sub, created=False):
```
Even if you don't use it, you still need to keep `parent` in the arg list, because it's a position argument. I hope that clears out the test failures, and then this will be ready to go.

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -404,11 +404,19 @@ class InstanceInstanceGroupsList(InstanceGroupMembershipMixin, SubListCreateAtta
     relationship = 'rampart_groups'
 
-    def is_valid_relation(self, parent, sub, created=False):
+    def is_valid_relation(self, parent, created=False):
```
I expect you'll need to add back in `sub` as well

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -408,7 +408,15 @@ def is_valid_relation(self, parent, sub, created=False):
+        return None
+
+    def is_valid_removal(self, parent, sub):
+        res = self.is_valid_relation(parent)
```
need to pass `parent, sub` here

AlanCoding reviewed (approved):
thanks for working through this, looks great.
